### PR TITLE
Optimize JDBC driver to be TP friendly

### PR DESCRIPTION
**Pull request by sunxiaoguang**
Existing JDBC driver was designed for BigData usage. This PR make it appropriate for TP workload.

close: #121

### CODE REVIEW

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/LoadBalancingDriver.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  private static final String URL_PROVIDER = "url.provider";
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String TIDB_MIN_DISCOVER_INTERVAL_KEY = "tidb.jdbc.min-discovery-interval";
+  private static final long TIDB_MIN_DISCOVER_INTERVAL = 1000;
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String TIDB_MAX_DISCOVER_INTERVAL_KEY = "tidb.jdbc.min-discovery-interval";
```
Comment (zhangyangyu) [inline jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/LoadBalancingDriver.java:65]: shouldn't the value be `tidb.jdbc.max-discovery-interval`?

Comment (sunxiaoguang) [reply jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/LoadBalancingDriver.java:65]: good catch!

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+package io.tidb.bigdata.jdbc.impl;
+
+import java.util.Arrays;
+import java.util.Random;
+import java.util.function.Function;
+
+public class RandomShuffleUrlMapper implements Function<String[], String[]> {
+
+  private static final ThreadLocal<Random> threadLocalRandom = ThreadLocal.withInitial(Random::new);
```
Comment (zhangyangyu) [inline jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java:25]: If we don't need to support Java 6 or older, we could directly use `java.util.concurrent.ThreadLocalRandom`.

**Review by humengyu2012 — COMMENTED** [review]
LGTM

**Review by humengyu2012 — APPROVED** [review]
LGTM

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RoundRobinUrlMapper.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+public class RoundRobinUrlMapper implements Function<String[], String[]> {
+
+  private final AtomicLong offset = new AtomicLong();
+
+  public String[] apply(final String[] input) {
+    int currentOffset = (int) (offset.getAndIncrement() % input.length);
+    if (currentOffset == 0 || currentOffset == input.length) {
```
Comment (zhangyangyu) [inline jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RoundRobinUrlMapper.java:28]: Seems not possible for `currentOffset == input.length`.

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java (hunk 2 above)
Comment (sunxiaoguang) [reply jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java:25]: Good point! We are not going to support versions older than 8.

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RoundRobinUrlMapper.java (hunk 3 above)
Comment (sunxiaoguang) [reply jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RoundRobinUrlMapper.java:28]: You are right. It's not going to happen.

File: jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java (hunk 2 above)
Comment (zhangyangyu) [reply jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/impl/RandomShuffleUrlMapper.java]: yes, we use lambdas.

**Review by zhangyangyu — APPROVED** [review]

**Review by humengyu2012 — APPROVED** [review]
LGTM

**Pull request merged**